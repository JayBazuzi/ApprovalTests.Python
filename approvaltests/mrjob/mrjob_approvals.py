from io import BytesIO
from itertools import product

from mrjob.job import MRJob
from approvaltests import verify, verify_all_combinations


def verify_map_reduce(mr_job_under_test: MRJob, test_data: str) -> None:
    storyboard = print_map_reduce_job(mr_job_under_test, test_data)
    verify(storyboard)


def print_map_reduce_job(mr_job_under_test: MRJob, test_data: str) -> str:
    storyboard = f"{test_data}\n\nMap reduces to:\n\n"
    mr_job_under_test.sandbox(stdin=BytesIO(test_data.encode("utf-8")))
    with mr_job_under_test.make_runner() as runner:
        runner.run()
        for key, value in mr_job_under_test.parse_output(runner.cat_output()):
            storyboard += f"{key}:{value}\n"
    return storyboard


def verify_templated_map_reduction(map_reduction, input_creator, params):
    inputs = product(*params)
    storyboard = ""
    for input in inputs:
        storyboard += f"===================\n\n{input} =>\n"
        data = input_creator(*input)
        storyboard += f"{print_map_reduce_job(map_reduction, data)}\n"
    verify(storyboard)
