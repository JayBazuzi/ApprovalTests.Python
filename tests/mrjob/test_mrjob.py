from mrjob.job import MRJob

from approvaltests.mrjob.mrjob_approvals import (
    verify_map_reduce,
    verify_map_reduction_for_combinations_of,
)


class MRWordFrequencyCount(MRJob):
    def mapper(self, _, line):
        yield "chars", len(line)
        yield "words", len(line.split())
        yield "lines", 1

    def reducer(self, key, values):
        yield key, sum(values)


def test_word_count():
    test_data = "one fish two fish red fish blue fish"
    map_reduction = MRWordFrequencyCount(["--no-conf"])
    verify_map_reduce(map_reduction, test_data)


def test_word_count_combinations():
    animals = ["fish"]
    # animals = ["fish", "dog", "cat"]
    colors = ["magenta", "chartreuse", "aqua", "red", "blue"]
    map_reduction = MRWordFrequencyCount(["--no-conf"])

    def input_creator(animal, color1, color2):
        return f"one {animal} two {animal} {color1} {animal} {color2} {animal}"

    verify_map_reduction_for_combinations_of(
        map_reduction, input_creator, [animals, colors, colors]
    )
