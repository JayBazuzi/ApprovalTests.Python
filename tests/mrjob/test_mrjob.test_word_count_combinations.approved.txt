===================

('fish', 'magenta', 'magenta') =>
one fish two fish magenta fish magenta fish

Map reduces to:

chars:43
lines:1
words:8

===================

('fish', 'magenta', 'chartreuse') =>
one fish two fish magenta fish chartreuse fish

Map reduces to:

chars:46
lines:1
words:8

===================

('fish', 'magenta', 'aqua') =>
one fish two fish magenta fish aqua fish

Map reduces to:

chars:40
lines:1
words:8

===================

('fish', 'magenta', 'red') =>
one fish two fish magenta fish red fish

Map reduces to:

chars:39
lines:1
words:8

===================

('fish', 'magenta', 'blue') =>
one fish two fish magenta fish blue fish

Map reduces to:

chars:40
lines:1
words:8

===================

('fish', 'chartreuse', 'magenta') =>
one fish two fish chartreuse fish magenta fish

Map reduces to:

chars:46
lines:1
words:8

===================

('fish', 'chartreuse', 'chartreuse') =>
one fish two fish chartreuse fish chartreuse fish

Map reduces to:

chars:49
lines:1
words:8

===================

('fish', 'chartreuse', 'aqua') =>
one fish two fish chartreuse fish aqua fish

Map reduces to:

chars:43
lines:1
words:8

===================

('fish', 'chartreuse', 'red') =>
one fish two fish chartreuse fish red fish

Map reduces to:

chars:42
lines:1
words:8

===================

('fish', 'chartreuse', 'blue') =>
one fish two fish chartreuse fish blue fish

Map reduces to:

chars:43
lines:1
words:8

===================

('fish', 'aqua', 'magenta') =>
one fish two fish aqua fish magenta fish

Map reduces to:

chars:40
lines:1
words:8

===================

('fish', 'aqua', 'chartreuse') =>
one fish two fish aqua fish chartreuse fish

Map reduces to:

chars:43
lines:1
words:8

===================

('fish', 'aqua', 'aqua') =>
one fish two fish aqua fish aqua fish

Map reduces to:

chars:37
lines:1
words:8

===================

('fish', 'aqua', 'red') =>
one fish two fish aqua fish red fish

Map reduces to:

chars:36
lines:1
words:8

===================

('fish', 'aqua', 'blue') =>
one fish two fish aqua fish blue fish

Map reduces to:

chars:37
lines:1
words:8

===================

('fish', 'red', 'magenta') =>
one fish two fish red fish magenta fish

Map reduces to:

chars:39
lines:1
words:8

===================

('fish', 'red', 'chartreuse') =>
one fish two fish red fish chartreuse fish

Map reduces to:

chars:42
lines:1
words:8

===================

('fish', 'red', 'aqua') =>
one fish two fish red fish aqua fish

Map reduces to:

chars:36
lines:1
words:8

===================

('fish', 'red', 'red') =>
one fish two fish red fish red fish

Map reduces to:

chars:35
lines:1
words:8

===================

('fish', 'red', 'blue') =>
one fish two fish red fish blue fish

Map reduces to:

chars:36
lines:1
words:8

===================

('fish', 'blue', 'magenta') =>
one fish two fish blue fish magenta fish

Map reduces to:

chars:40
lines:1
words:8

===================

('fish', 'blue', 'chartreuse') =>
one fish two fish blue fish chartreuse fish

Map reduces to:

chars:43
lines:1
words:8

===================

('fish', 'blue', 'aqua') =>
one fish two fish blue fish aqua fish

Map reduces to:

chars:37
lines:1
words:8

===================

('fish', 'blue', 'red') =>
one fish two fish blue fish red fish

Map reduces to:

chars:36
lines:1
words:8

===================

('fish', 'blue', 'blue') =>
one fish two fish blue fish blue fish

Map reduces to:

chars:37
lines:1
words:8

